openflow15/flowmod.go

```diff
@@ -0,0 +1,310 @@
+	n += 2
+	f.Importance = binary.BigEndian.Uint16(data[n:])
+	n += 2
+
+	f.Match.UnmarshalBinary(data[n:])
+	n += int(f.Match.Len())
+
+	for n < int(f.Header.Length) {
+		instr := DecodeInstr(data[n:])
+		f.Instructions = append(f.Instructions, instr)
+		n += int(instr.Len())
+	}
+	return nil
+}
+
+// ofp_flow_mod_command 1.5
+const (
+	FC_ADD = iota // OFPFC_ADD = 0
+	FC_MODIFY
+	FC_MODIFY_STRICT
+	FC_DELETE
+	FC_DELETE_STRICT
+)
+
+// ofp_flow_mod_flags 1.5
+const (
+	FF_SEND_FLOW_REM = 1 << 0 /* Send flow removed message when flow expires or is deleted. */
+	FF_CHECK_OVERLAP = 1 << 1 /* Check for overlapping entries first */
+	FF_RESET_COUNTS  = 1 << 2 /* Reset flow packet and byte counts */
+	FF_NO_PKT_COUNTS = 1 << 3 /* Don’t keep track of packet count */
+	FF_NO_BYT_COUNTS = 1 << 4 /* Don’t keep track of byte count */
+)
+
+// BEGIN: ofp13 - 7.4.2
+type FlowRemoved struct {
+	common.Header
+	TableId  uint8
+	Reason   uint8
+	Priority uint16
+
+	IdleTimeout uint16
+	HardTimeout uint16
+	Cookie      uint64
+
+	Match Match
+	Stats
+}
+
+func NewFlowRemoved() *FlowRemoved {
+	f := new(FlowRemoved)
+	f.Header = NewOfp15Header()
+	f.Header.Type = Type_FlowRemoved
+	f.Match = *NewMatch()
+	f.Stats = *NewStats() //OXS_OF_DURATION is mandatory
+	return f
+}
+
+func (f *FlowRemoved) Len() (n uint16) {
+	n = f.Header.Len()
+	n += 16
+	n += f.Match.Len()
+	n += f.Stats.Len()
+	return
+}
+
+func (f *FlowRemoved) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, int(f.Len()))
+	n := 0
+
+	f.Header.Length = f.Len()
+	bytes, err := f.Header.MarshalBinary()
+	if err != nil {
+		return
+	}
+	copy(data[n:], bytes)
+	n += int(f.Header.Len())
+
+	data[n] = f.TableId
+	n += 1
+
+	data[n] = f.Reason
+	n += 1
+
+	binary.BigEndian.PutUint16(data[n:], f.Priority)
+	n += 2
+
+	binary.BigEndian.PutUint16(data[n:], f.IdleTimeout)
+	n += 2
+	binary.BigEndian.PutUint16(data[n:], f.HardTimeout)
+	n += 2
+
+	binary.BigEndian.PutUint64(data[n:], f.Cookie)
+	n += 8
+
+	bytes, err = f.Match.MarshalBinary()
+	if err != nil {
+		return
+	}
+	copy(data[n:], bytes)
+	n += int(f.Match.Len())
+
+	var b []byte
+	b, err = f.Stats.MarshalBinary()
+	if err != nil {
+		return
+	}
+	copy(data[n:], b)
+	n += len(b)
+
+	return
+}
+
+func (f *FlowRemoved) UnmarshalBinary(data []byte) error {
+	n := 0
+	var err error
+	err = f.Header.UnmarshalBinary(data[n:])
+	if err != nil {
+		return err
+	}
+	n += int(f.Header.Len())
+
+	f.TableId = data[n]
+	n += 1
+	f.Reason = data[n]
+	n += 1
+	f.Priority = binary.BigEndian.Uint16(data[n:])
+	n += 2
+
+	f.IdleTimeout = binary.BigEndian.Uint16(data[n:])
+	n += 2
+	f.HardTimeout = binary.BigEndian.Uint16(data[n:])
+	n += 2
+
+	f.Cookie = binary.BigEndian.Uint64(data[n:])
+	n += 8
+
+	err = f.Match.UnmarshalBinary(data[n:])
+	if err != nil {
+		return err
+	}
+	n += int(f.Match.Len())
+
+	err = f.Stats.UnmarshalBinary(data[n:])
+	if err != nil {
+		return err
+	}
+	n += int(f.Stats.Len())
+
+	return err
```

Comment:
I generally prefer to return err so that in case the 'err' gets set, we return it.
I understand that in this particular function, we check for 'err != nil' and return immediately, but keeping 'return err' would make sure we don't accidentally return nil for an actual error.